### ISSUE TITLE
admin-all doesn't return all servers

### ISSUE THREAD

**Issue by acgustafson** [LOENS2/pterodactyl_exporter#23]
**Describe the bug**
Using admin-all returns all servers visible to the user, but not all servers visible in the pterodactyl panel. If you "share" a server with a user it is returned in the response from pterodactyl, but by default as an admin servers owned by other users and not shared with you are not reported by pterodactyl_exporter.

**To Reproduce**
```
import requests

url = "https://panel.myserver.com"
token = "secret"
headers = {"Authorization": f"Bearer {token}", "Content-Type": "application/json","Accept": "application/json"}

resp = requests.get(url + "/api/client", headers=headers).json()
for server in resp.get("data"):
    print(server.get("attributes").get("name"))
```

**Expected behavior**
All servers available to an admin including servers not shared with them would be visible as the name "admin-all" implies. 

I think the best solution here would be to add a flag specifying which API your token is for and then calling the client API or application API respectively. Calling the `/api/application/servers` API instead returns all servers as expected. That combined with updating the docs to describe that admin-all does not actually mean all servers in the panel, only servers owned by or shared with the user regardless of their admin status would help eliminate any future confusion.

**Comment by LOENS2** [comment]
That's interesting. I'll look into that at the weekend.

**Comment by LOENS2** [comment]
Would you please specify the problem further? For me, when using "admin-all" all servers from all users are shown to me by the exporter. This of course only applies if you have the admin role. The "" option provides the functionality you describe, only showing owned servers and servers you are added on as a user.
Please excuse my late reply. I'm quite busy at the moment.

**Comment by acgustafson** [comment]
I think I described it as best I can. If a different user owns the server and has not shared it with the admin it is not visible in the exporter. Perhaps there was a change at some point to the Pterodactyl API and we have different versions? I am running 1.11.5 which is the latest release from October of 2023. 

The config I have is
```
api_key: ptlc_*************************
host: ptero01.example.com
https: true
ignore_ssl: true
port: 9531
server_list_type: "admin-all"
```
In the Pterodactyl dashboard I see the following servers:
![image](https://github.com/LOENS2/pterodactyl_exporter/assets/10574673/8bdb8973-f669-46ec-8066-d4b6a2c9a0ed)

With the following server not owned by or shared with my user but still visible since it's an admin:
![image](https://github.com/LOENS2/pterodactyl_exporter/assets/10574673/22af8113-5bc8-410a-b30f-8c7359797eb2)

The exporter is only showing the following metrics:
```
# HELP python_gc_objects_collected_total Objects collected during gc
# TYPE python_gc_objects_collected_total counter
python_gc_objects_collected_total{generation="0"} 130.0
python_gc_objects_collected_total{generation="1"} 289.0
python_gc_objects_collected_total{generation="2"} 0.0
# HELP python_gc_objects_uncollectable_total Uncollectable object found during GC
# TYPE python_gc_objects_uncollectable_total counter
python_gc_objects_uncollectable_total{generation="0"} 0.0
python_gc_objects_uncollectable_total{generation="1"} 0.0
python_gc_objects_uncollectable_total{generation="2"} 0.0
# HELP python_gc_collections_total Number of times this generation was collected
# TYPE python_gc_collections_total counter
python_gc_collections_total{generation="0"} 45.0
python_gc_collections_total{generation="1"} 4.0
python_gc_collections_total{generation="2"} 0.0
# HELP python_info Python platform information
# TYPE python_info gauge
python_info{implementation="CPython",major="3",minor="10",patchlevel="7",version="3.10.7"} 1.0
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 2.5694208e+07
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 2.082816e+07
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.70887826037e+09
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 0.44999999999999996
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 7.0
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 1.073741816e+09
# HELP pterodacytl_server_memory_megabytes Memory used by server in megabytes
# TYPE pterodacytl_server_memory_megabytes gauge
pterodacytl_server_memory_megabytes{id="4ba7219c",server_name="test-satisfactory"} 0.0
pterodacytl_server_memory_megabytes{id="54da1554",server_name="matt-minecraft"} 12980.740096
pterodacytl_server_memory_megabytes{id="2c6dc441",server_name="palworld"} 0.0
# HELP pterodactyl_server_cpu_absolute Absolute cpu usage by server
# TYPE pterodactyl_server_cpu_absolute gauge
pterodactyl_server_cpu_absolute{id="4ba7219c",server_name="test-satisfactory"} 0.0
pterodactyl_server_cpu_absolute{id="54da1554",server_name="matt-minecraft"} 33.607
pterodactyl_server_cpu_absolute{id="2c6dc441",server_name="palworld"} 0.0
# HELP pterodactyl_server_disk_megabytes Disk space used by server in megabytes
# TYPE pterodactyl_server_disk_megabytes gauge
pterodactyl_server_disk_megabytes{id="4ba7219c",server_name="test-satisfactory"} 0.0
pterodactyl_server_disk_megabytes{id="54da1554",server_name="matt-minecraft"} 4982.910102
pterodactyl_server_disk_megabytes{id="2c6dc441",server_name="palworld"} 2676.588632
# HELP pterodactyl_server_network_rx_megabytes Megabytes received by server via network
# TYPE pterodactyl_server_network_rx_megabytes gauge
pterodactyl_server_network_rx_megabytes{id="4ba7219c",server_name="test-satisfactory"} 0.0
pterodactyl_server_network_rx_megabytes{id="54da1554",server_name="matt-minecraft"} 5803.770327
pterodactyl_server_network_rx_megabytes{id="2c6dc441",server_name="palworld"} 0.0
# HELP pterodactyl_server_network_tx_megabytes Megabytes transmitted by server via network
# TYPE pterodactyl_server_network_tx_megabytes gauge
pterodactyl_server_network_tx_megabytes{id="4ba7219c",server_name="test-satisfactory"} 0.0
pterodactyl_server_network_tx_megabytes{id="54da1554",server_name="matt-minecraft"} 50400.040378
pterodactyl_server_network_tx_megabytes{id="2c6dc441",server_name="palworld"} 0.0
# HELP pterodactyl_server_uptime_milliseconds Server uptime in milliseconds
# TYPE pterodactyl_server_uptime_milliseconds gauge
pterodactyl_server_uptime_milliseconds{id="4ba7219c",server_name="test-satisfactory"} 0.0
pterodactyl_server_uptime_milliseconds{id="54da1554",server_name="matt-minecraft"} 1.581189599e+09
pterodactyl_server_uptime_milliseconds{id="2c6dc441",server_name="palworld"} 0.0
# HELP pterodactyl_server_max_memory_megabytes Maximum memory allocated to server in megabytes
# TYPE pterodactyl_server_max_memory_megabytes gauge
pterodactyl_server_max_memory_megabytes{id="4ba7219c",server_name="test-satisfactory"} 0.0
pterodactyl_server_max_memory_megabytes{id="54da1554",server_name="matt-minecraft"} 14000.0
pterodactyl_server_max_memory_megabytes{id="2c6dc441",server_name="palworld"} 32768.0
# HELP pterodactyl_server_max_swap_megabytes Maximum swap allocated to server in megabytes
# TYPE pterodactyl_server_max_swap_megabytes gauge
pterodactyl_server_max_swap_megabytes{id="4ba7219c",server_name="test-satisfactory"} 0.0
pterodactyl_server_max_swap_megabytes{id="54da1554",server_name="matt-minecraft"} 0.0
pterodactyl_server_max_swap_megabytes{id="2c6dc441",server_name="palworld"} 0.0
# HELP pterodactyl_server_max_disk_megabytes Maximum disk space allocated to server in megabytes
# TYPE pterodactyl_server_max_disk_megabytes gauge
pterodactyl_server_max_disk_megabytes{id="4ba7219c",server_name="test-satisfactory"} 0.0
pterodactyl_server_max_disk_megabytes{id="54da1554",server_name="matt-minecraft"} 30000.0
pterodactyl_server_max_disk_megabytes{id="2c6dc441",server_name="palworld"} 20000.0
# HELP pterodactyl_server_io IO weight of server
# TYPE pterodactyl_server_io gauge
pterodactyl_server_io{id="4ba7219c",server_name="test-satisfactory"} 500.0
pterodactyl_server_io{id="54da1554",server_name="matt-minecraft"} 500.0
pterodactyl_server_io{id="2c6dc441",server_name="palworld"} 500.0
# HELP pterodactyl_server_max_cpu_absolute Maximum cpu load allowed to server
# TYPE pterodactyl_server_max_cpu_absolute gauge
pterodactyl_server_max_cpu_absolute{id="4ba7219c",server_name="test-satisfactory"} 0.0
pterodactyl_server_max_cpu_absolute{id="54da1554",server_name="matt-minecraft"} 0.0
pterodactyl_server_max_cpu_absolute{id="2c6dc441",server_name="palworld"} 0.0
```

If you share the "test-minecraft" server with the user then the exporter gets the details on the server.

**Comment by LOENS2** [comment]
Thank you for the further clarification. That is indeed a very strange problem. Especially, because I wasn't able to reproduce it within 6 hours of testing.
These are my results when using the exporter on my prod instance (Version 1.11.5):

An excerpt from my exporter output:
```
# HELP pterodactyl_server_disk_megabytes Disk space used by server in megabytes
# TYPE pterodactyl_server_disk_megabytes gauge
pterodactyl_server_disk_megabytes{id="63a2d375",server_name="Lobby"} 58.300592
pterodactyl_server_disk_megabytes{id="1f260d6c",server_name="Open Arena"} 417.537661
pterodactyl_server_disk_megabytes{id="d9c65259",server_name="Speedrun Random"} 327.27724
pterodactyl_server_disk_megabytes{id="999f7662",server_name="Velocity"} 61.472546
pterodactyl_server_disk_megabytes{id="231b0fd7",server_name="Counter Strike: Global Offensive"} 35095.428165
pterodactyl_server_disk_megabytes{id="79517706",server_name="Satisfactory"} 13235.775427
pterodactyl_server_disk_megabytes{id="8589ffd6",server_name="CSI Musik"} 66.206982
pterodactyl_server_disk_megabytes{id="1803cbb0",server_name="CSI Level und Co."} 22.206618
pterodactyl_server_disk_megabytes{id="ec26e910",server_name="Astropolis"} 616.882281
pterodactyl_server_disk_megabytes{id="18224307",server_name="The Silvans Project"} 558.650488
pterodactyl_server_disk_megabytes{id="a08a5bc4",server_name="Paul Musik Bot"} 33.119461
pterodactyl_server_disk_megabytes{id="492156d5",server_name="FiveM"} 427.812073
pterodactyl_server_disk_megabytes{id="efd35991",server_name="Euro Truck Simulator 2"} 617.71281
pterodactyl_server_disk_megabytes{id="ff289fa2",server_name="Paul"} 399.992495
pterodactyl_server_disk_megabytes{id="04e5ba8c",server_name="American Truck Simulator"} 575.602516
pterodactyl_server_disk_megabytes{id="e553f6bb",server_name="Bingo"} 291.848776
```

The Servers i own:

![image](https://github.com/LOENS2/pterodactyl_exporter/assets/72134951/2cecf451-ae48-41aa-9eb5-1ffbd02b78f9)

The others' servers:

![image](https://github.com/LOENS2/pterodactyl_exporter/assets/72134951/fd2e6610-ce8f-46c6-a346-bac59fca1f31)

BUT when using the `owner` option it get the following:

```
# HELP pterodactyl_server_disk_megabytes Disk space used by server in megabytes
# TYPE pterodactyl_server_disk_megabytes gauge
pterodactyl_server_disk_megabytes{id="63a2d375",server_name="Lobby"} 58.300592
pterodactyl_server_disk_megabytes{id="1f260d6c",server_name="Open Arena"} 417.537661
pterodactyl_server_disk_megabytes{id="999f7662",server_name="Velocity"} 61.472546
pterodactyl_server_disk_megabytes{id="231b0fd7",server_name="Counter Strike: Global Offensive"} 35095.428165
pterodactyl_server_disk_megabytes{id="79517706",server_name="Satisfactory"} 13235.775427
pterodactyl_server_disk_megabytes{id="ec26e910",server_name="Astropolis"} 616.882281
pterodactyl_server_disk_megabytes{id="18224307",server_name="The Silvans Project"} 558.650488
pterodactyl_server_disk_megabytes{id="492156d5",server_name="FiveM"} 427.812073
pterodactyl_server_disk_megabytes{id="efd35991",server_name="Euro Truck Simulator 2"} 617.713014
pterodactyl_server_disk_megabytes{id="04e5ba8c",server_name="American Truck Simulator"} 575.60272
pterodactyl_server_disk_megabytes{id="e553f6bb",server_name="Bingo"} 291.848781
```

As you may notice, the others' servers are missing in the second excerpt. This is what lets me believe, that there is some kind of config error. Just for completeness, here is my config:

```
host: 10.0.0.20
port: 9531
api_key: ptlc_[REDACTED]
https: false
ignore_ssl: true
server_list_type: "admin-all"
```

I wonder if this might be a problem with running in docker vs. running on bare metal. Would you please tell me if you run the exporter on bare metal or as a container. I couldn't find any differences between those versions, but that would be another potential point of failure.

**Comment by acgustafson** [comment]
That's very strange. I do use Docker but I don't think that affects it. If I run the python query I submitted above from the host I run the exporter on I  get this:
```
pterodactyl_exporter]# python3 test.py
test-satisfactory
matt-minecraft
palworld
```

Using the same API key and switching the API endpoint to `/api/application/servers` does give me all servers.
```
pterodactyl_exporter]# python3 test.py
test-satisfactory
test-minecraft
matt-minecraft
palworld
```

I tried generating a new API key and that doesn't change anything. I assume your Pterodactyl instance is a little older? Maybe this is something that changed and new instances are impacted but old ones have some old permission carried over from a prior version?

**Comment by LOENS2** [comment]
I'm still not able to recreate your problem. I'll reset my testing environment and start all over again. That is the weirdest problem I had so far.

**Comment by LOENS2** [comment]
Finally! I've managed to recreate the problem. This problem only occurs on the docker version of the exporter. I don't know why that is as I have to further investigate that. For a quick workaround I recommend using the exporter egg.

**Comment by LOENS2** [comment]
I've fixed the docker image! My build pipeline for docker has silently failed for the last year until today. Now the docker image should be on the latest version. The docker image you have doesn't even include the `admin-all` option and therefor just falls back to the `""` option. Thank you for bringing this to my attention! Please pull the latest image and check if it works. If so, please close the issue as completed.

**Comment by acgustafson** [comment]
Well there's a new problem now. Looks like you are trying to pass `check_hostname` to HTTPSConnection which isn't valid in the newer versions on http.client. 3.11 was the last version to support that in 3.12 it's no longer in the constructor. 

https://docs.python.org/3.12/library/http.client.html

```
]# docker compose up
[+] Running 10/10
 ✔ pterodactyl-exporter 9 layers [⣿⣿⣿⣿⣿⣿⣿⣿⣿]      0B/0B      Pulled                                                                                                                                                                                                                  7.6s    ✔ 4abcf2066143 Already exists                                                                                                                                                                                                                                                     0.0s    ✔ dca80dc46cec Already exists                                                                                                                                                                                                                                                     0.0s    ✔ fe9e15b6315c Pull complete                                                                                                                                                                                                                                                      0.5s    ✔ a8fd6f3f484f Pull complete                                                                                                                                                                                                                                                      0.2s    ✔ 4fc96b5c1ba4 Pull complete                                                                                                                                                                                                                                                      2.2s    ✔ d49f75afc828 Pull complete                                                                                                                                                                                                                                                      0.4s    ✔ 9cad514282fd Pull complete                                                                                                                                                                                                                                                      0.6s    ✔ 93a574f44e3d Pull complete                                                                                                                                                                                                                                                      1.1s    ✔ 0524e2d157d3 Pull complete                                                                                                                                                                                                                                                      1.0s [+] Running 2/2
 ✔ Network docker_default          Created                                                                                                                                                                                                                                           0.1s  ✔ Container pterodactyl-exporter  Created                                                                                                                                                                                                                                           0.3s Attaching to pterodactyl-exporter
pterodactyl-exporter  | Traceback (most recent call last):
pterodactyl-exporter  |   File "<frozen runpy>", line 198, in _run_module_as_main
pterodactyl-exporter  |   File "<frozen runpy>", line 88, in _run_code
pterodactyl-exporter  |   File "/opt/pterodactyl_exporter/app/pterodactyl_exporter/pterodactyl_exporter.py", line 48, in <module>
pterodactyl-exporter  |     main()
pterodactyl-exporter  |   File "/opt/pterodactyl_exporter/app/pterodactyl_exporter/pterodactyl_exporter.py", line 24, in main
pterodactyl-exporter  |     http_client.client_init(config)
pterodactyl-exporter  |   File "/opt/pterodactyl_exporter/app/pterodactyl_exporter/http_client.py", line 15, in client_init
pterodactyl-exporter  |     client = http.client.HTTPSConnection(config_file['host'], 443, check_hostname=not config_file['ignore_ssl'])
pterodactyl-exporter  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pterodactyl-exporter  | TypeError: HTTPSConnection.__init__() got an unexpected keyword argument 'check_hostname'
pterodactyl-exporter exited with code 1
[root@grafana docker]#
```

**Comment by LOENS2** [comment]
Should be fixed now with d9c16bb.

**Comment by acgustafson** [comment]
Looks like it's fixed now! Thanks
